apps/labs/posts/sparse-array-ecosystem.md

```diff
@@ -0,0 +1,215 @@
+
+The second main way we can take advantage of sparsity is via compute. By "skipping over" the background values; one can exceed the speed of a dense algorithm; once again assuming that the density does not exceed some threshold. This is, once again, not as simple as it may seem at first glance: A sparse algorithm needs to combine present values and their corresponding positions from different arrays and decide the value and presence of a corresponding element in the output array. We see a common pitfall here when writing sparse algorithms: If at any point; we accidentally iterate over the missing values, we risk losing many performance benefits of a sparse algorithm.
+
+### Common Formats
+
+In the last section, we hinted at different schemes for storing the positions of present values in sparse arrays. Such a scheme is known more commonly as a (sparse array) _format_. What format is best for storing some data is often determined by the pattern of present values in the data, the sparsity of the data, and some other considerations. Let's explore further by seeing a two common example formats for illustrative purposes.
+
+#### Dictionary-of-keys (DOK)
+
+This is one of the simplest formats to explain, so we begin with this one. The idea is that we have a dictionary that stores the coordinates of present values as the key, with the value representing the value of the array at that coordinate. Let's illustrate with a small example using the `sparse` library.
+
+```python
+>>> import numpy as np
+>>> import sparse
+>>> import pprint
+>>> eye_dok = sparse.DOK.from_numpy(np.eye(5))
+>>> pprint.pp(eye_dok.data)
+{(0, 0): 1.0, (1, 1): 1.0, (2, 2): 1.0, (3, 3): 1.0, (4, 4): 1.0}
+```
+
+This format has a number of advantages:
+
+1. It's simple to implement.
+2. It supports insertion into any coordinate by inserting the appropriate key-value pair into the dictionary.
+3. Random access is possible by retrieving a key from the dictionary.
+
+However, it also has a number of disadvantages:
+
+1. Iterating over values may be random, depending on the dictionary implementation. This creates problems when combining different arrays; and checking the presence of a given row or column (one can only check the presence of an element efficiently).
+2. It suffers from a higher memory usage than some other formats, as dictionaries take up more space than arrays.
+3. Retrieval of a single element may be fast, but it's slower than a similar access from an array due to hashing.
+
+### COOrdinate Format
+
+This format is also extremely simple, it stores one array of values, and a corresponding array of positions.
```

Comment:
```suggestion
This format is also relatively straightforward, it stores one array of values, and a corresponding array of positions.
```

Avoid telling the reader what is "simple". They may not agree, and, by telling them something is simple, you may stir feelings of deficiency in them for "not getting it".